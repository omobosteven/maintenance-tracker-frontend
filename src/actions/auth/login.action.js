import axios from 'axios';
import Cookie from 'cookies-js';
import jwt from 'jsonwebtoken';
import toastr from 'toastr';
import config from '../../config';
import routes from '../../constants/routes';
import {
  SET_CURRENT_USER, SET_CURRENT_USER_FAIL, DELETE_ERROR_MESSAGE
} from '../../constants/types';

export const setCurrentUser = user => ({
  type: SET_CURRENT_USER,
  user
});

export const setCurrentUserError = error => ({
  type: SET_CURRENT_USER_FAIL,
  error
});

export const deleteErrorMessages = () => ({
  type: DELETE_ERROR_MESSAGE
});

const loginAction = userDetails => dispatch => axios.post(`${config.apiUrl}${routes.SIGN_IN}`,
  userDetails)
  .then((response) => {
    const { message } = response.data;
    const { token } = response.data.data;
    Cookie.set('jwtToken', token);
    toastr.success(message);
    dispatch(setCurrentUser(jwt.decode(token)));
  })
  .catch((error) => {
    dispatch(setCurrentUserError(error.response.data));
  });

export default loginAction;
