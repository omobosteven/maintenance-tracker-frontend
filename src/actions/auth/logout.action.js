import Cookie from 'cookies-js';
import { LOGOUT_USER } from '../../constants/types';

export const logoutCurrentUser = () => ({
  type: LOGOUT_USER
});

const logoutAction = () => (dispatch) => {
  Cookie.expire('jwtToken');
  dispatch(logoutCurrentUser({}));
};

export default logoutAction;
