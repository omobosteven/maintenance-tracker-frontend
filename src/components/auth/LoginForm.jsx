import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import TextField from '../common/TextField';
import routes from '../../constants/routes';
import ValidateFormInputs from '../../validations/ValidateAuthInputs';
import ErrorAlertNotification from '../common/ErrorAlertNotification';
import loginAction, {
  deleteErrorMessages
} from '../../actions/auth/login.action';

export class LoginForm extends Component {
  state = {
    email: '',
    password: '',
    errors: {}
  }

  onChange = (event) => {
    const { errors } = this.state;
    if (errors[event.target.name]) {
      const newErrors = Object.assign({}, errors);
      delete newErrors[event.target.name];
      this.setState({
        [event.target.name]: event.target.value,
        errors: newErrors
      });
    } else {
      this.setState({
        [event.target.name]: event.target.value
      });
    }
  };

  onSubmit = (event) => {
    const { action } = this.props;
    event.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {} });
      action.loginUser(this.state);
    }
  };

  handleDelete = () => {
    const { action } = this.props;
    action.deleteError();
    this.setState({
      password: ''
    });
  };

  isValid = () => {
    const { errors, isValid } = ValidateFormInputs.loginInput(
      this.state
    );

    if (!isValid) {
      this.setState({ errors, password: '' });
    }

    return isValid;
  };

  render() {
    const { auth, error } = this.props;
    const {
      email, password, errors
    } = this.state;

    if (auth) {
      return <Redirect to={routes.USER_REQUESTS} />;
    }

    return (
      <section className="section-form container">
        {
          (error && error.message) ? (
            <ErrorAlertNotification
              id="errorNotify"
              errors={error.message}
              onClick={this.handleDelete}
            />
          ) : ''
        }
        <div className="row card-form">
          <form className="signup-form">
            <h2 className="form-title">User Login</h2>
            <TextField
              className="email"
              error={errors.email}
              label="Email"
              field="email"
              value={email}
              onChange={this.onChange}
              placeholder="Your email"
              type="email"
            />
            <TextField
              className="password"
              error={errors.password}
              label="Password"
              field="password"
              value={password}
              onChange={this.onChange}
              placeholder="Your password"
              type="password"
            />
            <div className="submit-btn">
              <input className="submit" type="submit" value="Login" onClick={this.onSubmit} />
              <p className="form-info">
                Not registered ? create an account
                <Link to={routes.SIGN_UP} className="form-link"> Signup</Link>
              </p>
            </div>
          </form>
        </div>
      </section>
    );
  }
}

LoginForm.propTypes = {
  auth: PropTypes.bool,
  error: PropTypes.instanceOf(Object),
  action: PropTypes.instanceOf(Object)
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  error: state.auth.error
});

const matchDispatchToProps = dispatch => ({
  action: bindActionCreators(
    {
      loginUser: loginAction,
      deleteError: deleteErrorMessages
    },
    dispatch
  )
});

export default connect(mapStateToProps, matchDispatchToProps)(LoginForm);
