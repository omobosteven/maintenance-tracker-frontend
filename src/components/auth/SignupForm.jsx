import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import TextField from '../common/TextField';
import routes from '../../constants/routes';
import ValidateFormInputs from '../../validations/ValidateAuthInputs';
import ErrorAlertNotification from '../common/ErrorAlertNotification';
import signupAction, {
  deleteErrorMessages
} from '../../actions/auth/signup.action';

export class SignupForm extends Component {
  state = {
    email: '',
    username: '',
    password: '',
    password_confirmation: '',
    errors: {}
  }

  onChange = (event) => {
    const { errors } = this.state;
    if (errors[event.target.name]) {
      const newErrors = Object.assign({}, errors);
      delete newErrors[event.target.name];
      this.setState({
        [event.target.name]: event.target.value,
        errors: newErrors
      });
    } else {
      this.setState({
        [event.target.name]: event.target.value
      });
    }
  };

  onSubmit = (event) => {
    const { action } = this.props;
    event.preventDefault();
    if (this.isValid()) {
      this.setState({ errors: {} });
      action.signupUser(this.state);
    }
  };

  handleDelete = () => {
    const { action } = this.props;
    action.deleteError();
    this.setState({
      password: '',
      password_confirmation: ''
    });
  };

  isValid = () => {
    const { errors, isValid } = ValidateFormInputs.signupInput(
      this.state
    );

    if (!isValid) {
      this.setState({ errors, password: '', password_confirmation: '' });
    }

    return isValid;
  };

  render() {
    const { auth, error } = this.props;
    const {
      username, email, password, password_confirmation, errors // eslint-disable-line
    } = this.state;

    if (auth) {
      return <Redirect to={routes.USER_REQUESTS} />;
    }

    return (
      <section className="section-form container">
        {
          (error && error.message) ? (
            <ErrorAlertNotification
              errors={error.message}
              onClick={this.handleDelete}
            />
          ) : ''
        }
        <div className="row card-form">
          <form className="signup-form">
            <h2 className="form-title">Create an account</h2>
            <TextField
              error={errors.username}
              label="Username"
              field="username"
              value={username}
              onChange={this.onChange}
              placeholder="Your username"
              type="text"
            />
            <TextField
              error={errors.email}
              label="Email"
              field="email"
              value={email}
              onChange={this.onChange}
              placeholder="Your email"
              type="email"
            />
            <TextField
              error={errors.password}
              label="Password"
              field="password"
              value={password}
              onChange={this.onChange}
              placeholder="Your password"
              type="password"
            />
            <TextField
              error={errors.password_confirmation}
              label="Confirm Password"
              field="password_confirmation"
              value={password_confirmation} // eslint-disable-line
              onChange={this.onChange}
              placeholder="Confirm your password"
              type="password"
            />
            <div className="submit-btn">
              <input type="submit" value="Create account" onClick={this.onSubmit} />
              <p className="form-info">
                Already have an account ?
                <Link to={routes.SIGN_IN} className="form-link"> Login</Link>
              </p>
            </div>
          </form>
        </div>
      </section>
    );
  }
}

SignupForm.propTypes = {
  auth: PropTypes.bool,
  error: PropTypes.instanceOf(Object),
  action: PropTypes.instanceOf(Object)
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  error: state.auth.error
});

const matchDispatchToProps = dispatch => ({
  action: bindActionCreators(
    {
      signupUser: signupAction,
      deleteError: deleteErrorMessages
    },
    dispatch
  )
});

export default connect(mapStateToProps, matchDispatchToProps)(SignupForm);
