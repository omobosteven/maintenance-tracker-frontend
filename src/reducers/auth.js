import isEmpty from 'is-empty';
import {
  SET_CURRENT_USER,
  DELETE_ERROR_MESSAGE,
  SET_CURRENT_USER_FAIL
} from '../constants/types';

const initialState = {
  isAuthenticated: false,
  user: {},
  error: {}
};

export default (state = initialState, action) => {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        isAuthenticated: !isEmpty(action.user),
        user: action.user
      };

    case SET_CURRENT_USER_FAIL:
      return {
        ...state,
        isAuthenticated: false,
        user: {},
        error: action.error
      };

    case DELETE_ERROR_MESSAGE:
      return {
        error: {}
      };

    default: return state;
  }
};
