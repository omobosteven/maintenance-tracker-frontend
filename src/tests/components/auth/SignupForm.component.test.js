import React from 'react';
import { shallow } from 'enzyme';
import { SignupForm } from '../../../components/auth/SignupForm';
import mockData from '../../__mocks__/mockData';

let wrapper;

describe('SignupForm', () => {
  beforeEach(() => {
    const props = {
      action: {
        signupUser: jest.fn(),
        deleteError: jest.fn()
      },
      error: {},
      auth: false,
    };
    wrapper = shallow(<SignupForm {...props} />);
  });

  it('renders the SignupForm component', () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should set username when username changes', () => {
    const event = {
      target: {
        name: 'username',
        value: 'blaze'
      }
    };

    const username = wrapper.find('.username');

    username.simulate('change', event);

    const expectedUsername = 'blaze';

    expect(wrapper.instance().state.username).toBe(expectedUsername);
  });

  it('should set password when password changes', () => {
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'password',
        value: 'password'
      }
    };

    const password = wrapper.find('.password');

    password.simulate('change', event);

    wrapper.setState({ errors: { [event.target.name]: ['mock'] } });
    wrapper.instance().onChange(event);
    expect(wrapper.instance().state[event.target.name])
      .toBe(event.target.value);
  });

  it('should Signup user when correct creadentials are supplied', () => {
    const { signUpData } = mockData;
    const event = {
      preventDefault: jest.fn(),
    };

    const Signup = wrapper.find('.submit');

    wrapper.setState(signUpData);
    Signup.simulate('click', event);

    expect(wrapper.instance().state.username).toBe(signUpData.username);
  });

  it('should return error for invalid form submission', () => {
    wrapper.find('.submit').simulate('click', {
      preventDefault: jest.fn(),
    });
    expect(wrapper.state('error')).not.toBe({});
    expect(wrapper).toMatchSnapshot();
  });

  it('should redirect user if user is already logged in', () => {
    const props = {
      action: {
        signupUser: jest.fn(),
        deleteError: jest.fn()
      },
      error: {},
      auth: true,
    };
    wrapper = shallow(<SignupForm {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should delete error message if there is an error', () => {
    const props = {
      action: {
        signupUser: jest.fn(),
        deleteError: jest.fn()
      },
      error: {
        message: 'There is an error'
      },
      auth: false,
    };
    wrapper = shallow(<SignupForm {...props} />);

    const deleteErrorMessage = wrapper.find('#errorNotify');
    deleteErrorMessage.simulate('click');

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.instance().state.password).toBe('');
  });
});
