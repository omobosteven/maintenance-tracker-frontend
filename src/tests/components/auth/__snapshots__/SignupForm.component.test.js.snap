// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupForm renders the SignupForm component 1`] = `
<section
  className="section-form container"
>
  <div
    className="row card-form"
  >
    <form
      className="signup-form"
    >
      <h2
        className="form-title"
      >
        Create an account
      </h2>
      <TextField
        className="username"
        field="username"
        label="Username"
        onChange={[Function]}
        placeholder="Your username"
        type="text"
        value=""
      />
      <TextField
        field="email"
        label="Email"
        onChange={[Function]}
        placeholder="Your email"
        type="email"
        value=""
      />
      <TextField
        className="password"
        field="password"
        label="Password"
        onChange={[Function]}
        placeholder="Your password"
        type="password"
        value=""
      />
      <TextField
        field="password_confirmation"
        label="Confirm Password"
        onChange={[Function]}
        placeholder="Confirm your password"
        type="password"
        value=""
      />
      <div
        className="submit-btn"
      >
        <input
          className="submit"
          onClick={[Function]}
          type="submit"
          value="Create account"
        />
        <p
          className="form-info"
        >
          Already have an account ?
          <Link
            className="form-link"
            replace={false}
            to="/auth/login"
          >
             Login
          </Link>
        </p>
      </div>
    </form>
  </div>
</section>
`;

exports[`SignupForm should delete error message if there is an error 1`] = `
<section
  className="section-form container"
>
  <ErrorAlertNotification
    errors="There is an error"
    id="errorNotify"
    onClick={[Function]}
  />
  <div
    className="row card-form"
  >
    <form
      className="signup-form"
    >
      <h2
        className="form-title"
      >
        Create an account
      </h2>
      <TextField
        className="username"
        field="username"
        label="Username"
        onChange={[Function]}
        placeholder="Your username"
        type="text"
        value=""
      />
      <TextField
        field="email"
        label="Email"
        onChange={[Function]}
        placeholder="Your email"
        type="email"
        value=""
      />
      <TextField
        className="password"
        field="password"
        label="Password"
        onChange={[Function]}
        placeholder="Your password"
        type="password"
        value=""
      />
      <TextField
        field="password_confirmation"
        label="Confirm Password"
        onChange={[Function]}
        placeholder="Confirm your password"
        type="password"
        value=""
      />
      <div
        className="submit-btn"
      >
        <input
          className="submit"
          onClick={[Function]}
          type="submit"
          value="Create account"
        />
        <p
          className="form-info"
        >
          Already have an account ?
          <Link
            className="form-link"
            replace={false}
            to="/auth/login"
          >
             Login
          </Link>
        </p>
      </div>
    </form>
  </div>
</section>
`;

exports[`SignupForm should redirect user if user is already logged in 1`] = `
<Redirect
  push={false}
  to="/users/requests"
/>
`;

exports[`SignupForm should return error for invalid form submission 1`] = `
<section
  className="section-form container"
>
  <div
    className="row card-form"
  >
    <form
      className="signup-form"
    >
      <h2
        className="form-title"
      >
        Create an account
      </h2>
      <TextField
        className="username"
        error={
          Array [
            "The username field is required.",
          ]
        }
        field="username"
        label="Username"
        onChange={[Function]}
        placeholder="Your username"
        type="text"
        value=""
      />
      <TextField
        error={
          Array [
            "The email field is required.",
          ]
        }
        field="email"
        label="Email"
        onChange={[Function]}
        placeholder="Your email"
        type="email"
        value=""
      />
      <TextField
        className="password"
        error={
          Array [
            "The password field is required.",
          ]
        }
        field="password"
        label="Password"
        onChange={[Function]}
        placeholder="Your password"
        type="password"
        value=""
      />
      <TextField
        error={
          Array [
            "The password confirmation field is required.",
          ]
        }
        field="password_confirmation"
        label="Confirm Password"
        onChange={[Function]}
        placeholder="Confirm your password"
        type="password"
        value=""
      />
      <div
        className="submit-btn"
      >
        <input
          className="submit"
          onClick={[Function]}
          type="submit"
          value="Create account"
        />
        <p
          className="form-info"
        >
          Already have an account ?
          <Link
            className="form-link"
            replace={false}
            to="/auth/login"
          >
             Login
          </Link>
        </p>
      </div>
    </form>
  </div>
</section>
`;
